New atomic reference counting algorithm

m-ou-se opened the pull request:
This implements a new 'wait-free' atomic reference counting algorithm based on the paper [Wait-Free Weak Reference Counting](https://www.microsoft.com/en-us/research/uploads/prod/2023/04/preprint-644bdf17da97d.pdf) by @mjp41, @sylvanc and @bensimner.

The paper contains a way to implement atomic reference counting *wait free*, for five operations:
1. Release weak (= Weak::drop)
2. Release strong (= Arc::drop)
3. Acquire weak (= Weak::clone and Arc::downgrade)
4. Acquire strong (= Arc::clone)
5. Acquire strong from weak (= Weak::upgrade)

`Weak::upgrade` must increase the strong count if it is nonzero. Unfortunately, processors do not have native 'increment if nonzero' instructions. Therefore, it is usually implemented with a CAS loop to increment the strong count only if it is not zero.

The paper shows a way to avoid this CAS loop to make it wait-free. By reserving the least significant bit in the strong counter (by shifting the counter one bit to the left), we can use that extra bit to indicate the final 'dropped' state in which the strong counter is permanently zero. Then `Weak::upgrade` can be implemented as a `fetch_add(2)`, which leaves the 'permanently zero' bit untouched.

This however does mean that Arc::drop must now do an additional operation. Not only must it decrement the strong counter, it must also use a CAS operation to set the 'permanently zero' bit if the counter is zero.

The paper also shows an optimized version of the algorithm in which an additional bit in the strong counter is reserved to indicate whether there have ever been any weak pointers. When this bit is not set, some steps can be skipped.

However, the algorithm from the paper is unfortunately not something we can directly use in for Rust's standard `Arc` and `Weak`, because we have more operations:

1. Weak::drop
2. Arc::drop
3. Weak::clone
4. Arc::clone
5. Weak::upgrade
7. Arc::downgrade
8. Arc::get_mut
9. Arc::try_unwrap
10. Arc::into_inner

Specifically `Arc::get_mut` requires locking the weak counter to temporarily block `Arc::downgrade` from completing. We cannot implement `Arc::downgrade` as just `Weak::clone`.

Our `Arc::downgrade` implementation is implemented as a CAS loop to increment the weak counter if it doesn't hold the special 'locked' (usize::MAX) value, similar to how Weak::upgrade uses a CAS loop (which is what the paper solves).

I have extended the algorithm by also reserving the lowest bit in the weak counter to represent the 'weak counter locked' state. That way, `Arc::downgrade` can be implemented as a `fetch_add(2)` rather than a CAS loop. It will still have to spin in case of a concurrent call to `Arc::get_mut`, but in absence of `Arc::get_mut`, this makes Arc and Weak wait-free.

The paper shows some promising benchmarking results. I have not benchmarked this change yet.

m-ou-se committed:
Use new atomic reference counting algorithm.

m-ou-se committed:
Fix whitespace.

m-ou-se committed:
Update debug printers for Arc.

m-ou-se commented:
I don't think rustc itself is a good benchmark, since it doesn't use `upgrade` and `downgrade` much. If anyone wants to help benchmark this, that is very much appreciated. Especially projects that use `Weak::upgrade` very often should see a significant improvement.

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/6312055655/job/17137292141) [(plain)](https://github.com/rust-lang/rust/commit/36a48f259541cf24cae1eb5c80b4497c062931f4/checks/17137292141/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Prepare all required actions
Getting action download info
Download action repository 'actions/checkout@v4' (SHA:8ade135a41bc03ea155e62e844d188df1ea18608)
Download action repository 'actions/upload-artifact@v3' (SHA:a8a3f3ad30e3422c9c7b888a15615d19a852ae32)
Complete job name: PR - mingw-check-tidy
git config --global core.autocrlf false
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
---
Removing intermediate container 597d1100a70a
 ---> ed0cd6c6d0ac
Step 6/10 : COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
 ---> 4c009f8f5391
Step 7/10 : RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt     && pip3 install virtualenv
Collecting binaryornot==0.4.4
  Downloading binaryornot-0.4.4-py2.py3-none-any.whl (9.0 kB)
Collecting boolean-py==4.0
  Downloading boolean.py-4.0-py3-none-any.whl (25 kB)
---
Building wheels for collected packages: reuse
  Building wheel for reuse (pyproject.toml): started
  Building wheel for reuse (pyproject.toml): finished with status 'done'
  Created wheel for reuse: filename=reuse-1.1.0-cp310-cp310-manylinux_2_35_x86_64.whl size=180117 sha256=2196c9034bf565528bbb1ee6dad4f753eb813f58822363e6b768f09c73e4d4ff
  Stored in directory: /tmp/pip-ephem-wheel-cache-vpgh6cnn/wheels/c2/3c/b9/1120c2ab4bd82694f7e6f0537dc5b9a085c13e2c69a8d0c76d
Installing collected packages: boolean-py, binaryornot, setuptools, reuse, python-debian, markupsafe, license-expression, jinja2, chardet
  Attempting uninstall: setuptools
    Found existing installation: setuptools 59.6.0
    Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
---
  Downloading virtualenv-20.24.5-py3-none-any.whl (3.7 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.7/3.7 MB 66.9 MB/s eta 0:00:00
Collecting filelock<4,>=3.12.2
  Downloading filelock-3.12.4-py3-none-any.whl (11 kB)
Collecting distlib<1,>=0.3.7
  Downloading distlib-0.3.7-py2.py3-none-any.whl (468 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 468.9/468.9 KB 126.6 MB/s eta 0:00:00
Collecting platformdirs<4,>=3.9.1
  Downloading platformdirs-3.10.0-py3-none-any.whl (17 kB)
Installing collected packages: distlib, platformdirs, filelock, virtualenv
Successfully installed distlib-0.3.7 filelock-3.12.4 platformdirs-3.10.0 virtualenv-20.24.5
Removing intermediate container 3b60f322912e
 ---> b5af011dc3a3
Step 8/10 : COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
 ---> 4e63b8b49746
 ---> 4e63b8b49746
Step 9/10 : COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/
 ---> 799dbccc3f6e
Step 10/10 : ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
Removing intermediate container 18d40394757d
 ---> 886e5b1c1b19
Successfully built 886e5b1c1b19
Successfully tagged rust-ci:latest
Successfully tagged rust-ci:latest
##[endgroup]
Built container sha256:886e5b1c1b195ce0d14c0ecc082d9d994c64a37a0dd873bc445ba71c139031fc
Uploading finished image sha256:886e5b1c1b195ce0d14c0ecc082d9d994c64a37a0dd873bc445ba71c139031fc to https://ci-caches.rust-lang.org/docker/31024a61e3c099341c4151411b61536c5a15e77991c1c58c8adf475bd61d8568f58aad59ac3149cca672919d01d664a203adb87f86665b006743ece22accf3f1
IMAGE          CREATED          CREATED BY                                      SIZE      COMMENT
886e5b1c1b19   1 second ago     /bin/sh -c #(nop)  ENV SCRIPT=TIDY_PRINT_DIF…   0B        
4e63b8b49746   2 seconds ago    /bin/sh -c #(nop) COPY file:078ea1d11e7b7cda…   367B      
b5af011dc3a3   4 seconds ago    |1 DEBIAN_FRONTEND=noninteractive /bin/sh -c…   23.9MB    
4c009f8f5391   10 seconds ago   /bin/sh -c #(nop) COPY file:ac591dd6bc5afa66…   5.33kB    
ed0cd6c6d0ac   11 seconds ago   |1 DEBIAN_FRONTEND=noninteractive /bin/sh -c…   23.1MB    
---
<missing>      5 weeks ago      /bin/sh -c #(nop)  LABEL org.opencontainers.…   0B        
<missing>      5 weeks ago      /bin/sh -c #(nop)  ARG LAUNCHPAD_BUILD_ARCH     0B        
<missing>      5 weeks ago      /bin/sh -c #(nop)  ARG RELEASE                  0B        

<botocore.awsrequest.AWSRequest object at 0x7f492402a290>
gzip: stdout: Broken pipe
xargs: docker: terminated by signal 13
https://ci-caches.rust-lang.org/docker/31024a61e3c099341c4151411b61536c5a15e77991c1c58c8adf475bd61d8568f58aad59ac3149cca672919d01d664a203adb87f86665b006743ece22accf3f1
sha256:886e5b1c1b195ce0d14c0ecc082d9d994c64a37a0dd873bc445ba71c139031fc
---
DirectMap4k:      171968 kB
DirectMap2M:     7168000 kB
DirectMap1G:    11534336 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
    Finished dev [unoptimized] target(s) in 0.03s
##[endgroup]
downloading https://ci-artifacts.rust-lang.org/rustc-builds-alt/d23062b5be6bbf5be15226c7f497c89cf7d3d6e1/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/llvm-d23062b5be6bbf5be15226c7f497c89cf7d3d6e1-true/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm
---
##[endgroup]
fmt check
tidy check
tidy: Skipping binary file check, read-only filesystem
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10'
Requirement already satisfied: pip in ./build/venv/lib/python3.10/site-packages (23.2.1)
Collecting black==23.3.0 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 7))
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.7/1.7 MB 46.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.7/1.7 MB 46.3 MB/s eta 0:00:00
Collecting click==8.1.3 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 34))
  Downloading click-8.1.3-py3-none-any.whl (96 kB)
Collecting importlib-metadata==6.7.0 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 38))
  Downloading importlib_metadata-6.7.0-py3-none-any.whl (22 kB)
  Downloading importlib_metadata-6.7.0-py3-none-any.whl (22 kB)
Collecting mypy-extensions==1.0.0 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 42))
  Downloading mypy_extensions-1.0.0-py3-none-any.whl (4.7 kB)
Collecting packaging==23.1 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 46))
  Downloading packaging-23.1-py3-none-any.whl (48 kB)
Collecting pathspec==0.11.1 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 50))
  Downloading pathspec-0.11.1-py3-none-any.whl (29 kB)
  Downloading pathspec-0.11.1-py3-none-any.whl (29 kB)
Collecting platformdirs==3.6.0 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 54))
  Downloading platformdirs-3.6.0-py3-none-any.whl (16 kB)
Collecting ruff==0.0.272 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 58))
  Downloading ruff-0.0.272-py3-none-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (5.9 MB)
Collecting tomli==2.0.1 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 77))
  Downloading tomli-2.0.1-py3-none-any.whl (12 kB)
Collecting typed-ast==1.5.4 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 81))
  Downloading typed_ast-1.5.4-cp310-cp310-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_12_x86_64.manylinux2010_x86_64.whl (877 kB)
  Downloading typed_ast-1.5.4-cp310-cp310-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_12_x86_64.manylinux2010_x86_64.whl (877 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 877.7/877.7 kB 155.0 MB/s eta 0:00:00
Collecting typing-extensions==4.6.3 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 107))
  Downloading typing_extensions-4.6.3-py3-none-any.whl (31 kB)
Collecting zipp==3.15.0 (from -r /checkout/src/tools/tidy/config/requirements.txt (line 114))
  Downloading zipp-3.15.0-py3-none-any.whl (6.8 kB)
Installing collected packages: zipp, typing-extensions, typed-ast, tomli, ruff, platformdirs, pathspec, packaging, mypy-extensions, click, importlib-metadata, black
Successfully installed black-23.3.0 click-8.1.3 importlib-metadata-6.7.0 mypy-extensions-1.0.0 packaging-23.1 pathspec-0.11.1 platformdirs-3.6.0 ruff-0.0.272 tomli-2.0.1 typed-ast-1.5.4 typing-extensions-4.6.3 zipp-3.15.0
linting python files
/checkout/src/etc/gdb_providers.py:174:30: E703 [*] Statement ends with an unnecessary semicolon
/checkout/src/etc/lldb_providers.py:649:36: E703 [*] Statement ends with an unnecessary semicolon
Found 2 errors.
[*] 2 potentially fixable with the --fix option.

python linting failed! Printing diff suggestions:
--- /checkout/src/etc/gdb_providers.py
+++ /checkout/src/etc/gdb_providers.py
@@ -171,7 +171,7 @@
         self.weak = self.ptr["weak"]["v" if is_atomic else "value"]["value"]
         if is_atomic:
             if self.weak % 2 == 1:
-                self.weak = 0;
+                self.weak = 0
             elif self.weak > 0:
                 self.weak >>= 1
                 self.weak -= 1
--- /checkout/src/etc/lldb_providers.py
+++ /checkout/src/etc/lldb_providers.py
+++ /checkout/src/etc/lldb_providers.py
@@ -646,7 +646,7 @@
         self.weak_count = self.weak.GetValueAsUnsigned()
         if self.is_atomic:
             if self.weak_count % 2 == 1:
-                self.weak_count = 0;
+                self.weak_count = 0
             elif self.weak_count > 0:
                 self.weak_count >>= 1
                 self.weak_count -= 1
Would fix 2 errors.
some tidy checks failed
some tidy checks failed
tidy error: checks with external tool 'ruff' failed
  local time: Tue Sep 26 11:40:11 UTC 2023
  network time: Tue, 26 Sep 2023 11:40:12 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/6312128080/job/17137531674) [(plain)](https://github.com/rust-lang/rust/commit/819d5311593cc45df6e80dbefbe988d85e897a33/checks/17137531674/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
running 398 tests
........................................................................................  88/398
........................................................................................ 176/398
........................................................................................ 264/398
malloc(): unaligned fastbin chunk detected
error: test failed, to rerun pass `-p alloc --lib`
Caused by:
Caused by:
  process didn't exit successfully: `/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/release/deps/alloc-877a6e9443107475 --quiet -Z unstable-options --format json` (signal: 6, SIGABRT: process abort signal)
...................................................................................  local time: Tue Sep 26 11:52:48 UTC 2023
  network time: Tue, 26 Sep 2023 11:52:48 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

m-ou-se committed:
Fix debug assert.

mjp41 commented:
Awesome. Super excited to see how this performs outside of the microbenchmarks in our paper.

Please feel free to reach out if you want help.

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit 34ed37ca21e3f8b5d372551e3243eef9f0ed0b6f with merge 21047641b533575599dedac7012b881d0d7cce1b...
<!-- homu: {"type":"TryBuildStarted","head_sha":"34ed37ca21e3f8b5d372551e3243eef9f0ed0b6f","merge_sha":"21047641b533575599dedac7012b881d0d7cce1b"} -->

m-ou-se commented:
Once the `try` build is done, there is a toolchain that can be installed with [`rustup-toolchain-install-master`](https://github.com/kennytm/rustup-toolchain-install-master) to try out this new implementation.

The rust-timer bot will also run some benchmarks on rustc itself, but those are unlikely to show anything interesting, because rustc doesn't have any meaningful usage of Weak::upgrade.

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/6314431046/job/17147414907)
Build commit: 21047641b533575599dedac7012b881d0d7cce1b (`21047641b533575599dedac7012b881d0d7cce1b`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/6314431046/job/17147414907"},"merge_sha":"21047641b533575599dedac7012b881d0d7cce1b"} -->

rust-timer commented:
Queued 21047641b533575599dedac7012b881d0d7cce1b with parent 8bf0dec101dc458a411401a826994006abd2f022, future [comparison URL](https://perf.rust-lang.org/compare.html?start=8bf0dec101dc458a411401a826994006abd2f022&end=21047641b533575599dedac7012b881d0d7cce1b).
There is currently 1 preceding artifact in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~2.3 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

mjp41 commented on library/alloc/src/sync.rs:
```diff
@@ -2112,55 +2158,48 @@ impl<T: Clone, A: Allocator + Clone> Arc<T, A> {
+            {
+                // This is the only Arc, but other Weak pointers might exist.
+                // Weak pointers can no longer upgrade at this point, though.
+                if this.inner().weak.load(Relaxed) == ONE_WEAK {
+                    // No Weak pointers exist.
+                    this.inner().strong.store(ONE_STRONG_WITH_WEAK, Release); // Restore the strong count for this Arc.
```
At this point would it be safe to return to the state `(ONE_STRONG, NO_WEAK)`?

m-ou-se replied:
Yes, I think so!

mjp41 commented on library/alloc/src/sync.rs:
```diff
@@ -2313,26 +2352,23 @@ impl<T: ?Sized, A: Allocator> Arc<T, A> {
+                if self
+                    .inner()
+                    .weak
+                    .compare_exchange(NO_STRONG_WITH_WEAK, STRONG_DROPPED, Acquire, Relaxed)
```
This line seems to be using incorrect constants.
```suggestion
                    .compare_exchange(ONE_WEAK, NO_WEAK, Acquire, Relaxed)
```
Though, I guess the `weak` has the lock bit set, so maybe it is something else?

m-ou-se replied:
Oh oops. This should have been `ONE_WEAK` and `WEAK_LOCKED`. They have the same values though, so this doesn't change correctness, it was just very confusing. ^^'

mjp41 commented on library/alloc/src/sync.rs:
```diff
@@ -1561,35 +1587,27 @@ impl<T: ?Sized, A: Allocator> Arc<T, A> {
+                    // in the strong counter, and also add the implicit weak pointer to the weak
+                    // pointer counter.
+                    this.inner().weak.fetch_add(ONE_WEAK, Relaxed);
+                    this.inner().strong.fetch_or(NO_STRONG_WITH_WEAK, Relaxed);
```
This flag setting can be an `add` as it will only be done once.

m-ou-se replied:
Yeah, but I am guessing that there might be more architectures with an atomic bit set instruction than an atomic add instruction.

mjp41 replied:
In the paper, we just used `fetch_add` as it reduces the number of primitives that needed to be wait-free for the algorithm to be wait-free. But in practice, I'm sure you're right that a platform that provides a wait-free `fetch_add` will provide an atomic bit set.

mjp41 reviewed (changes requested):
@m-ou-se this is really cool to see.  Mostly minor, but I think `downgrade` is incorrect.  Please let me know, if you want more explanation of the +2, it is quite subtle, and not covered well in our paper.

m-ou-se commented:
Thanks for your detailed review! I will try to address your comments tomorrow.

m-ou-se committed:
Fix comment in Arc impl

Co-authored-by: Matthew Parkinson <mjp41@users.noreply.github.com>

m-ou-se commented on library/alloc/src/sync.rs:
```diff
@@ -1561,35 +1587,27 @@ impl<T: ?Sized, A: Allocator> Arc<T, A> {
-                cur = this.inner().weak.load(Relaxed);
-                continue;
-            }
+            let prev = this.inner().weak.fetch_add(ONE_WEAK, Acquire);
```
> In the case where this is locked, you will loop and re-execute this line, which will add multiple weak references.

When the low bit of the weak pointer is set, the rest of the bits are meaningless (just like in the strong counter in your paper). So it doesn't matter that fetch_add is repeated, as long as it left the least significant bit set.

> Also, the paper checks if it is weak == 0 and if it observes this adds two. This is really important, and not part of your implementation. Without that sequence, it is possible to reach weak == 0 multiple times, which is very bad.

My code does that, but as a separate step: a few lines below here, in the `if prev == 0`, I add an extra `ONE_WEAK`, so we raise the weak counter by two in total if it was zero.

mjp41 replied:
> > In the case where this is locked, you will loop and re-execute this line, which will add multiple weak references.
> 
> When the low bit of the weak pointer is set, the rest of the bits are meaningless (just like in the strong counter in your paper). So it doesn't matter that fetch_add is repeated, as long as it leaves the least significant bit set.

Nice. That's great. I hadn't got that bit. Thanks for explaining.

> > Also, the paper checks if it is weak == 0 and if it observes this adds two. This is really important, and not part of your implementation. Without that sequence, it is possible to reach weak == 0 multiple times, which is very bad.
> 
> My code does that, but as a separate step: a few lines below here, in the `if prev == 0`, I add an extra `ONE_WEAK`, so we raise the weak counter by two in total if it was zero.

Sorry I see you increase by two, but doing in two increments was my concern. I was convinced it had to be done in one go, but now I'm trying to write out the counter example and can't find one. I am now convinced it is okay.

At some point, I'll try to extend @bensimner's proof to this variation.

m-ou-se committed:
Use the correct constants in Arc impl.

The values are the same, but I used the wrong constants.
So, this doesn't change any behaviour. It was just very confusing to
read.

m-ou-se committed:
Reset back to 'no weak' state in Arc::get_mut.

rust-timer commented:
Finished benchmarking commit (21047641b533575599dedac7012b881d0d7cce1b): [comparison URL](https://perf.rust-lang.org/compare.html?start=8bf0dec101dc458a411401a826994006abd2f022&end=21047641b533575599dedac7012b881d0d7cce1b&stat=instructions:u).

### Overall result: ❌ regressions - ACTION NEEDED


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=8bf0dec101dc458a411401a826994006abd2f022&end=21047641b533575599dedac7012b881d0d7cce1b&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.7% | [0.2%, 1.8%] | 5     |
| Regressions ❌ <br /> (secondary)  | 0.7% | [0.4%, 1.6%] | 7     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 0.7% | [0.2%, 1.8%] | 5     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=8bf0dec101dc458a411401a826994006abd2f022&end=21047641b533575599dedac7012b881d0d7cce1b&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 5.1%  | [0.1%, 10.1%]  | 2     |
| Regressions ❌ <br /> (secondary)  | 3.3%  | [3.3%, 3.3%]   | 1     |
| Improvements ✅ <br /> (primary)   | -5.7% | [-9.7%, -2.7%] | 4     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -2.1% | [-9.7%, 10.1%] | 6     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=8bf0dec101dc458a411401a826994006abd2f022&end=21047641b533575599dedac7012b881d0d7cce1b&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 1.6% | [1.6%, 1.6%] | 1     |
| Regressions ❌ <br /> (secondary)  | 1.2% | [1.2%, 1.2%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 1.6% | [1.6%, 1.6%] | 1     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=8bf0dec101dc458a411401a826994006abd2f022&end=21047641b533575599dedac7012b881d0d7cce1b&stat=size:linked_artifact)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.2%  | [0.0%, 0.6%]   | 47    |
| Regressions ❌ <br /> (secondary)  | 0.2%  | [0.2%, 0.2%]   | 1     |
| Improvements ✅ <br /> (primary)   | -0.1% | [-0.1%, -0.1%] | 2     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | 0.2%  | [-0.1%, 0.6%]  | 49    |
</details>

**Bootstrap**: 631.784s -> 629.276s (-0.40%)
**Artifact size**: 317.18 MiB -> 317.32 MiB (0.04%)

m-ou-se commented:
Which overhead are you referring to? Doesn't this already take a fast path basically everywhere for the 'no weak' case?

terrarier2111 commented:
Maybe this concern is completely irrational but i would assume because of the frequent usage of weakless Arcs the additional bit ops (which are just 1 additional cycle per strong decrement) could add up to outweigh the benefits from improving the weak impl, but maybe i am just wrong about that

Kobzol commented:
Maybe we could try to benchmark this with the parallel frontend enabled (I'm not sure if it actually uses Arc heavily though). CC @SparrowLii.

terrarier2111 commented:
Nvm i saw now that that op can just be applied at compile time (probably). I was talking about the shift in the drop part, if the optimizer understands that it can just modify the RHS then there won't be any additional ops needed. (the only additional thing on the fastpath with simple new + clone + drop operations will be the mov for the argument to drop_slow but that is probably negligible as it will only be executed once the Arc is dropped.

SparrowLii commented:
> Maybe we could try to benchmark this with the parallel frontend enabled (I'm not sure if it actually uses Arc heavily though). CC @SparrowLii.

So far, I have not found that Arc has a significant effect on the efficiency of parallel compilation.
https://github.com/rust-lang/rust/pull/109480#issuecomment-1480492465

m-ou-se committed:
Fix comment.

m-ou-se committed:
Properly reset weak count in Arc::make_mut.

m-ou-se commented:
> I was talking about the shift in the drop part, if the optimizer understands that it can just modify the RHS of the comparison instead of actually applying the shift to the LHS value then there won't be any additional ops needed.

Yeah, that compiles down to a single comparison just fine: https://godbolt.org/z/TvMTq5fez

(I wrote it as a shift because I think that shows the intent more clearly.)

mjp41 commented on library/alloc/src/sync.rs:
```diff
@@ -2112,55 +2158,48 @@ impl<T: Clone, A: Allocator + Clone> Arc<T, A> {
+                // Weak pointers can no longer upgrade at this point, though.
+                if this.inner().weak.load(Relaxed) == ONE_WEAK {
+                    // No Weak pointers exist.
+                    this.inner().strong.store(ONE_STRONG, Release); // Restore the strong count for this Arc.
```
I don't think you need the `compare_exchange`.  You can just use an assignment inside the `if`.

On line 2176, you know that STRONG_DROPPED and ONE_WEAK, this means the current Arc is the only one, there are no `weak`, so no one else can be inspecting the state.  So you are free to write them both out in either order.

What you have written is correct.  Personally, I prefer to only use a CAS is strictly required.

terrarier2111 commented:
> > I was talking about the shift in the drop part, if the optimizer understands that it can just modify the RHS of the comparison instead of actually applying the shift to the LHS value then there won't be any additional ops needed.
> 
> Yeah, that compiles down to a single comparison just fine: https://godbolt.org/z/TvMTq5fez
> 
> (I wrote it as a shift because I think that shows the intent more clearly.)

Alright, sorry for the confusion 😅

bors commented:
:umbrella: The latest upstream changes (presumably #115546) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

mjp41 commented:
@m-ou-se out of interest what is the bar for this making it into Rust?  Do you need an application that shows a win, or just no one showing a lose?

m-ou-se commented:
@mjp41 I think we should continue to merge if we can find some real world (or at least realistic) code that has a clear win. If we only show no performance regressions but no improvements, I don't think we should merge it. This change makes it a bit harder to maintain the Arc code for future maintainers, so it has to be worth it.

@yijunyu sent me a list of all crates on crates.io that call `Weak::upgrade`:

<details><summary>List</summary>

flattiverse_connector-36.1.1
melodium-0.5.1
druid-shell-0.7.0
zenoh-router-0.5.0-beta.7
zenoh-0.6.0-beta.1
basalt-0.18.0
ntex-0.5.27
rspack_style-0.1.16
mediasoup-0.11.0
wayland-scanner-0.30.0-beta.12
gst-plugin-threadshare-0.8.4
bevy_mod_scripting-0.1.2
rspack_style1-0.1.2
exocore-store-0.1.24
crndm-0.1.0
sodium-rust-2.1.1
kompact-0.11.0
smithay-0.3.0
winit-0.27.4
medea-0.2.0
caldera-0.2.0
luminvent_winit-0.26.1
nut-0.1.1
tauri-winit-0.24.1
stardust-xr-fusion-0.17.5
azul-winit-0.24.0
netidx-0.12.1
snocat-0.6.0-alpha.12
async-stream-packed-0.2.2
notation_model-0.5.0
zbox-0.9.2
ciphercore-base-0.1.2
anthill-di-1.2.4
condure-1.7.0
aeron-rs-0.1.3
device_query-1.1.1
webrtc-0.5.1
requiem-http-1.0.1-r1
scrappy-http-0.0.1
actori-http-1.0.1
openaws-vpn-client-0.1.4
ate-1.3.0
wlambda-0.8.1
cart-tmp-winit-0.22.2
anode-0.1.0
tox_core-0.1.1
actix-0.13.0
unbase-0.0.2
drumbeat-0.1.1
fuzzcheck-0.12.1
eternal-0.3.2
makiko-0.2.0
quickjs_runtime-0.8.5
async-coap-0.1.0
all-is-cubes-0.4.0
geobacter-runtime-core-1.0.0
glock-0.1.2
rustpython-vm-0.1.2
rclrust-0.0.2
rucene-0.1.1
lightning-signer-core-0.1.0-5
glib-0.16.0
rafx-framework-0.0.15
netxserver-1.7.3
weak-list2-0.1.0
congee-0.2.16
rust_engineio-0.4.0
woodchipper-1.1.0
tantivy-nightly-0.17.0-202205251639
corundum-0.4.1
fluvio-cluster-0.7.1
tetsy-jsonrpc-http-server-15.1.0
assemble-core-0.1.2
enfipy-jsonrpc-http-server-15.0.0
authority-round-0.1.0
rust-libcore-0.0.3
es_runtime-0.1.4
jsonrpc-http-server-18.0.0
tantivy-0.18.1
summavy-0.19.0
sentinel-core-0.1.2
gst-plugin-togglerecord-0.8.0
con-art-rust-0.2.0
oxigraph-0.3.6
futures-signals-0.3.31
async-slot-0.1.0
xi-core-lib-0.3.0
tetsy-updater-1.12.0
steamworks-0.9.0
frappe-0.4.7
tc-network-0.8.0
librice-0.0.2
wasmer-vm-near-2.4.0
xtor-0.9.10
async-acme-0.3.1
xiod-0.14.1
stack_test_epic_servers-3.0.3
broker-tokio-0.2.16
mugle_servers-5.2.0-alpha.5
fluence-fork-libp2p-0.36.2
examples-0.4.0
vulkan-malloc-0.1.5
tetsy-libp2p-0.34.3
tc-btree-0.7.0
rafx-resources-0.0.2
vapcore-1.12.1
syslog-rs-0.4.1
tet-libp2p-0.34.0
epic_servers-3.0.0
tokio-udt-0.1.0-alpha.8
grin_servers-5.1.2
crazyflie-lib-0.1.1
crs-bind-0.1.5
sgrankin-tacho-0.5.1
zee-0.3.2
agui_core-0.3.0
sc-network-0.9.0
libp2p-0.49.0
futures-executor-preview-0.3.0-alpha.19
nikidb-0.1.1
tacho-0.4.2
jujube-lib-0.1.1
remote-trait-object-0.5.0
ensync-1.0.1
webwire-0.4.0
graphsync-0.1.0
dependent_view-1.0.2
scrappy-client-0.0.1
requiem-wc-1.0.1-r1
tor-guardmgr-0.7.0
actoriwc-1.0.1
chamomile-0.8.0
awc-3.0.1
fluence-fork-libp2p-core-0.27.2
concurrency_traits-0.7.2
wayland-backend-0.1.0-beta.12
abel-0.1.1
mongodb_cwal-0.6.7
btleplug-0.10.1
everscale-network-0.3.12
tetsy-libp2p-core-0.27.1
carboxyl-0.2.1
bflog-0.3.1
incinerator-0.0.1
tet-libp2p-core-0.27.0
futures-executor-0.3.25
async_event_streams-0.1.4
wayland-server-0.30.0-beta.12
hakuban-0.6.1
sqlx-core-guts-0.6.0
local-pool-with-id-0.1.1
ipipe-0.11.7
bruteforus-0.1.0
single_executor-0.4.1
sqlx-core-0.6.2
indicatif-0.17.1
xiod_fakedata-0.4.1
prodash-21.0.0
fumio-reactor-0.1.0
hyper_wasi-0.15.0
dokan-0.2.0+dokan150
xtra-0.5.2
safe_cell_exts-0.3.0
polyhorn-core-0.4.0
uflow-0.6.1
hyper-0.14.20
cyfs-bdt-0.6.5
vex-rt-0.11.1
libutp-rs-1.0.0
observe-0.1.0
libp2p-helper-0.4.0
flash_rust_ws-0.4.1
hreq-h1-0.3.10
fluence-fork-libp2p-mplex-0.27.2
cell_rc-0.2.0
actix-web-4.2.1
dittolive-ditto-2.0.7
private-tx-1.0.0
tetsy-libp2p-mplex-0.27.2
shared_arena-0.8.4
sharedptr-0.3.4
fclones-0.29.1
appro-eq-0.3.1
hirofa_utils-0.5.5
ergo-rest-0.5.0
zenoh-collections-0.6.0-beta.1
drc-0.1.2
fluvio-wasm-timer-0.2.5
flatland-0.3.3
quickwit-metastore-0.3.0
mplex-0.27.0
futures-timer-3.0.2
rxrs-0.2.0-beta3
vrust-0.0.1
minimum-0.1.0
skywalking-0.4.0
wasm-timer-0.2.5
maia-0.1.1
kayrx-0.18.0
gw2lib-3.0.0-alpha-2
tokio_wasi-1.21.2
amethyst_assets-0.15.3
salvo_extra-0.37.1
nannou_wgpu-0.18.0
eventum-0.1.1
ydb-0.4.2
bottle-1.1.0-alpha
sardonyx_assets-0.0.3
vapcore-secretstore-1.0.0
fuel-p2p-0.11.2
tokio-sync-0.1.8
slack-morphism-1.3.2
phabricator-0.0.4
zbus-3.2.0
bleasy-0.2.2
verneuil-0.6.4
rdom-0.2.0
aspartam-0.1.0
tokio-1.21.2
sgx_tstd-1.1.1
delix-0.2.4
jojo-core-0.1.0
e-nguyen-0.1.2
parking_lot_mpsc-0.1.5
newport_gpu-0.2.0
vapcore-clique-0.1.0
madsim-0.2.8
rtrtr-0.2.2
ibc-relayer-cli-1.0.0
supernova-0.5.0
dipstick-0.9.0
healslut-0.1.0
hidamari-0.1.0
noders-0.0.2
elvis-core-0.1.1
libp2p-blake-streams-0.1.1
pipitor-0.3.0-alpha.15
bb8-0.8.0
tf_observer-0.1.2
futures-util-0.3.25
webrtc-ice-0.8.1
twitch-irc-5.0.0
lucet-runtime-internals-0.6.1
embedded-svc-0.22.1
arendur-0.0.5
dvcompute_branch-1.3.7
aptos-executor-0.2.7
netxclient-1.7.3
dvcompute_dist-1.3.7
dvcompute_cons-1.3.5
nash-native-client-0.3.0
git-branchless-lib-0.5.0
netmod-tcp-0.4.0
chronobreak-0.1.0
khronos-egl-4.1.0
tentacle-0.4.1
event-listener-primitives-2.0.1
timely-0.12.0
safe_core-0.43.1
dvcompute-1.3.4
bonsaidb-local-0.4.1
lettre-0.10.1
zenoh-transport-0.6.0-beta.1
metaplex-pulsar-4.1.1
madsim-tokio-postgres-0.2.0
futures-channel-0.3.25
dioxus-liveview-0.1.0
simple_futures-0.1.2
volo-thrift-0.2.0
rustable-0.3.0
zenoh-link-udp-0.6.0-beta.1
tracing-subscriber-0.3.16
crossfire-0.1.7
tokio-core-0.1.18
substrate-wasmtime-0.19.0
dynamic-pooling-1.0.0
ratchet_core-0.2.0
moka-0.9.4
ezk-sip-core-0.1.1
yukikaze-1.0.10
crymap-1.0.0
sn-pulsar-4.1.3
kayrx-timer-0.1.1
jack-0.10.0
cdbc-pg-0.1.22
scrappy-actor-0.0.1
mun_memory-0.2.0
cdbc-mysql-0.1.22
tokio-timer-0.2.13
tokio-postgres-0.7.7
tokio-tasker-1.2.0
sage_broker-0.3.0
requiem-framed-0.3.0-r1
rbdc-pg-0.1.19
linux-aio-tokio-0.3.0
tinychain-0.11.0
scrappy-framed-0.0.1
pulsar-4.1.3
rbdc-mysql-0.1.17
lock_api-0.4.9
ketos-0.12.0
actori-framed-0.3.0
keclc-framed-0.1.0
fumio-pool-0.1.0
actix-framed-0.3.1
threads_pool-0.2.6
dyn-cache-0.12.2
safina-executor-0.3.3
async-rustbus-0.1.2
udbg-0.2.1
requiem-0.9.0
nimiq-utils-0.2.0
ibc-relayer-0.19.0
asim-0.1.0
actori-0.9.0
potatonet-node-0.4.3
kpgres-0.5.0
zara-1.0.7
pkgcraft-0.0.2
fallacy-arc-0.1.1
ump-0.9.0
rcell-2.0.0-pre0
r2d2-0.8.10
nearly_eq-0.2.4
evc-0.1.2
finchers-0.13.5
daab-0.4.0
alto-3.0.4
futures-util-preview-0.3.0-alpha.19
ezsockets-0.3.0
webrtc-connection-0.2.0
ipfs-0.2.1
locutus-core-0.0.2
render_readme-0.7.5
abi_stable-0.10.4
libp2p-gossipsub-0.42.1
ckb-cli-1.1.1
linux-support-0.0.25
tokio-graceful-shutdown-0.11.1
minidsp-daemon-0.1.4
actix-raft-0.4.4
napi-2.10.0
tiny_http_sccache-0.7.0
txrx-0.1.0
gothack-future-parking_lot-0.3.4
tiny_http-0.12.0
async-liveliness-monitor-0.1.0
mdbook-pdf-headless_chrome-0.1.0
android-activity-0.4.0-beta.1
signal-stack-0.1.0
exonum-explorer-service-1.0.0
msr-core-0.3.5
sarekt-0.0.4
nakadion-0.30.0
janus_core-0.1.17
rouille-maint-in-3.0.1
future-parking_lot-0.3.3
casper-node-1.4.8
trillium-server-common-0.3.0
gotham-0.7.1
v-clickhouse-rs-0.2.0-alpha.7
streamunordered-0.5.2
interledger-store-redis-0.4.0
epoxy_streams-0.3.1
tetsy-hash-fetch-1.12.0
termusic-0.7.4
sophon-wasm-0.18.1
nobs-vulkanism-headless-0.1.0
cyfs-task-manager-0.6.0
cosync-0.2.1
watchexec-2.0.2
clickhouse-rs-1.0.0-alpha.1
rouille-ng-3.0.1
phoenix_channels_client-0.1.0
cozal-0.0.2
cosmian-wit-bindgen-rust-0.1.1
tokio-net-0.2.0-alpha.6
shredder-0.2.0
opentelemetry_sdk-0.18.0
anthill-service-system-1.2.3
ipfs-embed-0.24.0
heph-rt-0.4.1
wayland-cursor-0.30.0-beta.12
rouille-3.6.1
headless_chrome-0.9.0
conjure-runtime-4.0.0
tokio-threadpool-0.1.18
datafusion-13.0.0
amethyst_rendy-0.15.3
agner-reg-0.3.11
hey_listen-0.5.0
futures-locks-pre-0.5.1-pre
dwindow-0.3.0
ciruela-0.6.12
web-glitz-0.3.0
vmnet-0.1.1
virtiofsd-1.4.0
cachepot-0.1.0-rc.1
zerogc-context-0.2.0-alpha.7
tc-client-api-2.0.0
stack_test_epic_api-3.0.3
silver-rs-0.2.0-dev
sc-client-api-3.0.0
romio-0.3.0-alpha.10
akashi-0.5.2
yrs-warp-0.2.0
tubez-0.0.1
swindon-0.7.8
paxakos-0.12.0
async-web-server-0.6.2
async-task-ffi-4.1.1
vhdl_lang-0.17.0
vapcore-light-1.12.0
tube-0.0.2
rppal_w_frontend-0.0.4
reredis-0.1.0-alpha.2
mugle_api-5.2.0-alpha.5
kate-0.1.0
imdl-indicatif-0.14.0
eventuals-0.6.7
darksteel-0.1.0
bluest-0.5.3
abel-core-0.1.1
pi_async-0.5.5
netperf-0.2.3
moleculer-0.3.5
cargo-web-0.6.26
terminus-store-0.19.9
poem-1.3.47
output-0.6.2
futures-net-0.6.0
accesskit_windows-0.6.1
vrp-core-1.19.0
tracing-mutex-0.2.1
termwiz-0.18.0
syndicate-0.23.0
static_locks-0.1.0
salak_factory-0.10.0
roslibrust-0.5.1
lightproc-0.3.6-alpha.0
barrage-0.2.3
async-task-4.3.0
vapcore-sync-1.12.0
unmp-link-udp-0.7.0
unit-rs-0.2.0
rpi_embedded-0.1.0
rodio_wav_fix-0.15.0
joycon-rs-0.6.3
routing-0.37.1
madsim-etcd-client-0.2.8
futures-0.3.25
flo-state-1.1.0
crabquery-0.1.9
async-ucx-0.1.1
tokio-reactor-0.1.12
futures-rate-0.1.5
bandsocks-runtime-0.1.0
async-ws-0.3.3
validator-set-0.1.0
panorama-imap-0.0.4
libevent-0.1.0
gfx-backend-gl-0.9.0
cxx-async-0.1.1
sccache-0.3.0
reqwest_wasi-0.11.12
reqwest-wasm-0.11.15
nimiq-rpc-server-0.2.0
nails-0.13.0
jrsonnet-gcmodule-0.3.4
esp-idf-svc-0.42.5
dominator-0.5.31
byte_channel-0.0.1
aria2-ws-0.3.0
salvia-0.1.0
mz_rusoto_core-0.46.0
mwapi-0.4.1
grin_api-5.1.2
epic_api-3.0.0
douglas-0.1.1
vulkano-0.31.1
ticketed_lock-0.3.0
switchyard-0.3.0
rppal-0.13.1
azul-webrender-0.62.2
agner-actors-0.3.11
trade-0.1.0
pomfrit-0.1.9
gstreamer-0.18.8
dirinventory-0.7.0
asset_lru-0.1.3
stakker-0.2.5
rusoto_core-0.48.0
rodio-0.16.0
web_worker-0.3.0
vgtk-0.3.0
spin-0.9.4
spdlog-rs-0.2.4
reqwest-0.11.12
phabricator-mock-0.0.4
meli-0.7.2
gomoku-0.1.0
winit-modular-0.1.1
rsdb-0.12.1
curl-0.4.44
shared_lru-0.1.5
froggy-0.4.2
crayfish-0.0.1
blunt-0.0.8
xxlib-0.3.3
smartcalc-tui-1.0.8
minibus-0.0.2
wasmyon-0.1.1
hjul-0.2.2
stm-core-0.4.0
heng_rs-0.1.0
rw_lease-0.1.0
</details>

So, all crates whose performance might be increased by this PR should all be in this list. It'd be useful to find one or more crates in this list that heavily rely on Weak::upgrade in a hot path.

Another option would be to write some (realistic?) benchmarks that we can include in the standard library's test suite and/or [the runtime benchmark suite of rustc-perf](https://github.com/rust-lang/rustc-perf/tree/master/collector/runtime-benchmarks).

oskgo commented:
Ping from triage:
@m-ou-se Are you working on the benchmarks? If not, maybe opening an issue could help increase visibility?

JohnCSimon commented:
@m-ou-se 

ping from triage - can you post your status on this PR? There hasn't been an update in a few year, but I don't want to close it because it looks quite important.
